app/bundles/EmailBundle/Tests/Controller/Api/EmailApiControllerFunctionalTest.php

```diff
@@ -203,4 +219,232 @@ public function testReplyAction(): void
+        $this->assertEquals($sendResponse, ['success' => true, 'sentCount' => 1, 'failedRecipients' => 0], $clientResponse->getContent());
+
+        $testEmailOwnerAsMailer = function (): void {
+            $message = $this->transport->sentMessage;
+            $this->assertSame($message->getSubject(), 'Email created via API test');
+            $bodyRegExp = '#<h1>Email content created by an API test</h1><img height="1" width="1" src="[^"]+" alt="" />#';
+            $this->assertMatchesRegularExpression($bodyRegExp, $message->getBody());
+            $this->assertSame($message->getTo(), ['jane@api.test' => 'Jane Doe']);
+            $this->assertSame($message->getFrom(), ['john@api.test' => 'John Doe']);
+            $this->assertSame($message->getReplyTo(), ['john@api.test' => null]);
+            $this->assertSame($message->getBcc(), ['bcc@api.test' => null]);
+        };
+        $testEmailOwnerAsMailer();
+
+        // Send to contact:
+        $this->client->request('POST', "/api/emails/${emailId}/contact/${contactId}/send");
+        $clientResponse = $this->client->getResponse();
+
+        $this->assertSame(200, $clientResponse->getStatusCode(), $clientResponse->getContent());
+
+        $sendResponse   = json_decode($clientResponse->getContent(), true);
+
+        $this->assertEquals($sendResponse, ['success' => true], $clientResponse->getContent());
+        $testEmailOwnerAsMailer();
+
+        // Delete:
+        $this->client->request('DELETE', "/api/emails/{$emailId}/delete");
+        $clientResponse = $this->client->getResponse();
+        $response       = json_decode($clientResponse->getContent(), true);
+
+        $this->assertSame(200, $clientResponse->getStatusCode(), $clientResponse->getContent());
+
+        // Delete segment:
+        $this->client->request('DELETE', "/api/segments/{$segmentId}/delete");
+        $clientResponse = $this->client->getResponse();
+        $response       = json_decode($clientResponse->getContent(), true);
+
+        $this->assertSame(200, $clientResponse->getStatusCode(), $clientResponse->getContent());
+
+        // Delete contact:
+        $this->client->request('DELETE', "/api/contacts/{$contactId}/delete");
+        $clientResponse = $this->client->getResponse();
+        $response       = json_decode($clientResponse->getContent(), true);
+
+        $this->assertSame(200, $clientResponse->getStatusCode(), $clientResponse->getContent());
+
+        // Delete user:
+        $this->client->request('DELETE', "/api/users/{$userId}/delete");
+        $clientResponse = $this->client->getResponse();
+        $response       = json_decode($clientResponse->getContent(), true);
+
+        $this->assertSame(200, $clientResponse->getStatusCode(), $clientResponse->getContent());
+
+        // Clear owners cache (to leave a clean environment for future tests):
+        $mailHelper = self::$container->get('mautic.helper.mailer');
+        $this->setPrivateProperty($mailHelper, 'leadOwners', []);
```

Comment:
@fedys as our resident phpunit expert, is this cleanup needed or will the `tearDown()` method take care of this?

Reply:
There is no need for running 'DELETE' endpoints. As @escopecz wrote, the `tearDown()` method will take care of that.

The following block:
```php
// Clear owners cache (to leave a clean environment for future tests):
$mailHelper = self::$container->get('mautic.helper.mailer');
$this->setPrivateProperty($mailHelper, 'leadOwners', []);
```
should be moved to `EmailApiControllerFunctionalTest.php::tearDown()` method to make sure it is called in case the test fails.

FYI, `Clear owners cache (to leave a clean environment for future tests)` will no longer be needed once we ditch this line: https://github.com/mautic/mautic/blob/4.x/app/bundles/CoreBundle/Test/AbstractMauticTestCase.php#L62

Reply:
Thank you for your valuable suggestions. I have removed the HTTP `DELETE` code, since you mentioned that `MauticMysqlTestCase::tearDown` takes care of it. And I moved the cache cleanup to a `tearDown` method on this class.